core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -677,8 +677,13 @@ public SqlNode toSql(RexProgram program, RexNode rex) {
         }
         if (op instanceof SqlBinaryOperator && nodeList.size() > 2) {
           // In RexNode trees, OR and AND have any number of children;
-          // SqlCall requires exactly 2. So, convert to a left-deep binary tree.
-          return createLeftCall(op, nodeList);
+          // SqlCall requires exactly 2. So, convert to a balanced binary
+          // tree for OR/AND, left-deep binary tree for others.
+          if (op.kind == SqlKind.OR || op.kind == SqlKind.AND) {
+            return createBalancedCall(op, nodeList, 0, nodeList.size());
```

Comment:
Can we set a threshold for balanced call ? The left call looks more intuitive in the plan although it may causes StackOverflow.

Reply:
Nah, I don't think it is necessary.